src/btree/bt_curnext.c

```diff
@@ -756,36 +756,30 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
     need_walk = false;
     session = CUR2S(cbt);
     total_skipped = 0;
+
+    WT_STAT_CONN_DATA_INCR(session, cursor_next);
+
+    flags = WT_READ_NO_SPLIT | WT_READ_SKIP_INTL; /* tree walk flags */
+    if (truncating)
+        LF_SET(WT_READ_TRUNCATE);
+
+    WT_ERR(__wt_cursor_func_init(cbt, false));
+
     /*
      * If we have a bound set we should position our cursor appropriately if it isn't already
-     * positioned.
-     *
-     * In one scenario this function returns and needs to walk to the next record. In that case we
-     * set a boolean. We need to be careful that the entry flag isn't set so we don't re-enter
-     * search near.
-     *
-     * This logic must come before the call to __wt_cursor_func_init, as it will set the cbt active
-     * flag which changes the logic flow within the subsequent search near resulting in a
-     * segmentation fault. It is better to not set state prior to calling search near. Additionally
-     * the cursor initialization function does take a snapshot if required as does cursor search
-     * near.
+     * positioned. In one scenario this function returns and needs to walk to the next record. In
+     * that case we set a boolean.
      */
-    if (F_ISSET(cursor, WT_CURSTD_BOUND_LOWER) && !WT_CURSOR_IS_POSITIONED(cbt) &&
-      !F_ISSET(cursor, WT_CURSTD_BOUND_ENTRY)) {
+    if (F_ISSET(cursor, WT_CURSTD_BOUND_LOWER) && !WT_CURSOR_IS_POSITIONED(cbt)) {
         WT_ERR(__wt_btcur_bounds_row_position(session, cbt, true, &need_walk));
-        if (!need_walk)
+        if (!need_walk) {
+            __wt_value_return(cbt, cbt->upd_value);
             return (0);
+        }
     }
-    WT_STAT_CONN_DATA_INCR(session, cursor_next);
-
-    flags = WT_READ_NO_SPLIT | WT_READ_SKIP_INTL; /* tree walk flags */
-    if (truncating)
-        LF_SET(WT_READ_TRUNCATE);
 
     F_CLR(cursor, WT_CURSTD_KEY_SET | WT_CURSTD_VALUE_SET);
 
-    WT_ERR(__wt_cursor_func_init(cbt, false));
```

Comment:
Moving this looks like it will meaningfully change semantics for next. Was that all intentional?

For example: It used to be that __cursor_reset was called here (as part of cursor_func_init), which would reset any context left from btcur_bounds_row_position.

It also doesn't feel quite right to split calling cursor_func_init apart from clearing the cursor position at line 781. They feel like a pair of operations to me.

Reply:
Good point! From my understanding __wt_cursor_func_init only resets the cursor in if argument was set to **true**, in the case that it is set to false, we do not reset the cursor. Because we have changed btcur_row_position to use cursor_row_search(), a cursor_func_init() must be called before the search is being done, to set up the cursor correctly.

In regards to clearing the cursor flags here
` F_CLR(cursor, WT_CURSTD_KEY_SET | WT_CURSTD_VALUE_SET);`. How I understand this functionality is that clearing this means we want to find a key and value from cursor. After `__wt_btcur_bounds_row_position`, the flags will be set, because we have returned a key and value. The key and value returned here is not the key and value we actually want to return, unless we don't need to walk the tree due to **need_walk** being true.